### PR TITLE
Feat/remove team locking

### PR DESCRIPTION
**Pull request by TeddyRoncin**
## Before

Players had to lock their team by hand. They couldn't unlock.

## Now

The lock is automatic, it is done whenever a team is full and all the members have paied. There is also a waiting list system, that allows that, when a place is freed, teams which were locked first (but could not be accepted in the tournament because their wasn't enough place) automatically have take that place.

This means that unlocking is also automatic : if a player leaves a team, this team is unlocked (and can let its place to another team in the waiting list).

There are multiple situations in which we may need to lock or unlock teams :
* If a player pays its basket, we have to verify each player ticket one by one, and if necessary lock the team(s) of players whose their place is in the basket (or place the teams in the waiting list). It's not only the team of the user, because we could imagine that user quit team between the moment of the creation of the basket and the callback from the bank to confirm the payment.
* If we force-pay the place of a player, we need to verify if we can lock its team.
* Si we refund a basket, we need to look at every player tickets, and unlock the teams of these players.
* Si a player leaves a team, we have to unlock it.
* If a player is kicked of a team, we have to unlock it.
* If a player arrives in a team, if he fills the last player slot, and if its place is paid (by him or someone else), we may need to lock the team (we need to verify that each member of the team has a place before).
* If a team is unlocked and is not in the waiting list, we remove the team that has been in the waiting list for the longest, and give it a place in the tournament.
* If a team is disbanded, we unlock it.
* If a team in a solo tournament is created, and that the player creating the team has already a place for this tournament, we have to lock the team

There is a schema to sum that up :
Schéma team locking drawio
(sorry, it's in french, i don't want to translate it :/)

### CODE REVIEW

File: src/operations/carts.ts [hunk 1]
```diff
@@ -157,6 +191,18 @@ export const forcePay = async (user: User) => {
     }
   }
 
+  // First, verify the user is a player and that he has a team
+  if (user.type === UserType.player && team) {
+    // Check whether every player in the team has paid (except for this one, but we know it's going to be paid just after this)
+    const hasEveryonePaied = team.players.every((player) => player.id === user.id || player.hasPaid);
```
Comment (DevNono) [inline src/operations/carts.ts:197]: Typo => hasEveryonePaied

File: src/operations/carts.ts [hunk 2]
```diff
@@ -157,6 +191,18 @@ export const forcePay = async (user: User) => {
...
+    const tournament = await fetchTournament(team.tournamentId);
+    // Verify the team is full
+    const isTeamFull = team.players.length === tournament.playersPerTeam;
+    if (hasEveryonePaied && isTeamFull) {
```
Comment (DevNono) [inline src/operations/carts.ts:201]: Même typo

File: src/operations/team.ts [hunk 3]
```diff
@@ -24,6 +26,13 @@ const teamInclusions = {
   },
 };
 
+export const getPositionInQueue = (team: Team): Promise<number | undefined> => {
+  if (!team.enteredQueueAt) return undefined;
```
Comment (DevNono) [inline src/operations/team.ts:30]: Il faudrait mieux `return null`

File: tests/etupay/callback.test.ts [hunk 4]
```diff
@@ -48,58 +53,161 @@ const createEtupayPayload = (etupayBody: object) => {
...
+    lolTeam = await createFakeTeam({ members: lolTournament.playersPerTeam, tournament: 'lol' });
+    // Fill the tournament
+    for (let index = 0; index < lolTournament.placesLeft; index++) {
+      promises.push(createFakeTeam({ members: lolTournament.playersPerTeam, tournament: 'lol', locked: true }));
+    }
+    // Fetch the first 2 players, and force pay the others
+    [lolPlayer1, lolPlayer2] = lolTeam.players;
+    for (const player of lolTeam.players.slice(2)) await cartOperations.forcePay(player);
+
+    const csgoTournament = await tournamentOperations.fetchTournament('cs2');
```
Comment (DevNono) [inline tests/etupay/callback.test.ts:149]: Ca doit être renommé cs2 maintenant

**Review by DevNono — CHANGES_REQUESTED** [review]
Quelques changements mineurs à faire avant le merge

File: src/operations/carts.ts (hunk 1 above)
Comment (TeddyRoncin) [reply src/operations/carts.ts:197]: oh

**Review by DevNono — APPROVED** [review]

**Pull request merged**